### PR TITLE
Update the dockerfile to build faster and fixed dockerignore

### PR DESCRIPTION
**Pull request by schultzp2020**
Signed-off-by: Paul Schultz <pschultz@pobox.com>

## What does this PR do / why we need it

- Updates dockerfile to build faster
- Fixes dockerignore
- Adds a pr template

## Which issue(s) does this PR fix

N/A

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,65 +1,62 @@
...
 RUN tar xzf bundle.tar.gz && rm bundle.tar.gz
 
 # Copy any other files that we need at runtime
-COPY app-config.yaml ./
-
-RUN fix-permissions ./
+COPY ./app-config.yaml .
 
-CMD ["node", "packages/backend", "--config", "app-config.yaml"]
+CMD ["node", "packages/backend", "--config", "app-config.yaml"]
```
Comment (tumido) [inline Dockerfile:62]: Missing EOL at EOF

File: Dockerfile [hunk 2]
```diff
@@ -1,65 +1,62 @@
...
 USER 1001
 
 # Copy the install dependencies from the build stage and context
 COPY --from=build /opt/app-root/src/yarn.lock /opt/app-root/src/package.json /opt/app-root/src/packages/backend/dist/skeleton.tar.gz ./
 RUN tar xzf skeleton.tar.gz && rm skeleton.tar.gz
 
-RUN npm install -g yarn && \
-    yarn install --frozen-lockfile --production --network-timeout 600000 && rm -rf "$(yarn cache dir)"
+# Install production dependencies
+RUN yarn install --frozen-lockfile --production && yarn cache clean
```
Comment (tumido) [inline Dockerfile:53]: Not sure why this was changed. As you can see in the [PR check run](https://github.com/janus-idp/redhat-backstage-build/actions/runs/4019272506/jobs/6905903518), the `--network-timeout` is necessary.

**Review by tumido — CHANGES_REQUESTED** [review]
I like the changes in general. Besides the necessary change pointed out in the inline comment, this LGTM.

Out of curiosity, can you explain what makes the build faster here? Is it just the generic reduction of layer count (from 34 to 32) and removal of permission changes (which results in copy of all the modified files in the new layer) or is there something more to it?

**Comment by schultzp2020** [comment]
> Out of curiosity, can you explain what makes the build faster here? Is it just the generic reduction of layer count (from 34 to 32) and removal of permission changes (which results in copy of all the modified files in the new layer) or is there something more to it?

1. Recursively changing permissions on a folder can take a long time. Specially this can take a really long time with `node_modules`. Most of the time an image user has no need to have permissions.
2. Separating the install stage from the build stage helps with caching. i.e. A change in the source code that does not change `node_modules` will cache and skip the first stage.

File: Dockerfile (hunk 2 above)
Comment (schultzp2020) [reply Dockerfile:53]: Huh, I can change this back, I don't understand why this is an issue. On my local machine, the entire image build takes only 3 minutes.

**Comment by schultzp2020** [comment]
@tumido any idea why one of the status checks is failing?

File: Dockerfile (hunk 2 above)
Comment (tumido) [reply Dockerfile]: Yeah I think it's totally specific to GitHub Action environments only. We've also discovered by trial and error, when our builds failed in GH Actions but went just fine in other environments..

Anyways the network timeout fortunately has no negative impact on local builds so I guess it's fine.🙂

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -25,23 +34,26 @@ jobs:
...
-      - name: checkout repo
-        uses: actions/checkout@755da8c3cf115ac066823e79a1e1788f8940201b # pin@v3
+      - name: Checkout Repository
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
 
       - name: Set up QEMU
-        uses: docker/setup-qemu-action@e81a89b1732b9c48d79cd809d8d81d79c4647a18 #pin@v2.1.0
+        uses: docker/setup-qemu-action@v2
```
Comment (tumido) [inline .github/workflows/ci.yaml:43]: Please don't undo work of others #9 

GH action version pinning to hash is here for a reason. I'm not sure how unpinning them is making the build any faster (as per the PR title). It definitely makes it less secure.

https://michaelheap.com/ensure-github-actions-pinned-sha/
https://exercism.org/docs/building/github/gha-best-practices#h-pin-actions-to-shas

File: .github/workflows/docker-build.yaml [hunk 4]
```diff
@@ -73,16 +84,17 @@ jobs:
     name: process-image-manifest
     needs: ["setup", "build"]
     steps:
-      - name: Log in to Registry
-        uses: redhat-actions/podman-login@632d91dfe19e1b55833cb708786bfbad2c2a0335 #pin@v3.2.0
+      - name: Log in to Quay.io
```
Comment (tumido) [inline .github/workflows/docker-build.yaml:87]: We are not using quay yet. We want to support multiple registries and so far we support only GHCR. See #13

**Review by tumido — CHANGES_REQUESTED** [review]
Please revert your changes to GH workflow or reduce them to minimal. Due to few reasons:

- Atomic PRs are always better than changing the whole repository around in a single PR. Your `.editorconfig` and PR issue template changes now depend on a docker file changes - not ideal. This PR is getting out of hands with you modifying many different and unrelated things at once.
- Current CI setup ensures different workflows for `push` and `pull_request` events.  The `push` workflow already contains all the content of the `pull_request` one with addition of pushing the resulting image, we currently have no intention of pushing on PRs. The only way we may want to change that is potentially to use shared workflows/actions. Your current change makes it so that on `push` event all containers are built twice - first time in the `ci.yaml` workflow and second time in the `docker-build.yaml` workflow. Why?
- Rewamping whole CI code organization within a PR that is not focused on this change and without any discussion on that change - as a kind of an addendum to Docker file changes is not community friendly.

**Comment by sabre1041** [comment]
@schultzp2020 the use of the `fix-permissions` script is important when operating in environments that dynamically use a random UID at runtime, such as OpenShift. The upstream backstage image does not account for this and it causes the container to fail at runtime.

If we can determine if any directories can be omitted from permission modification without causing failures, that could be an improvement

File: .gitattributes [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# source code
```
Comment (tumido) [inline .gitattributes:1]: Can you please explain why is this file necessary here? What does it change? Do we expect PHP or XML files in this repo? It seems like a boilerplate for the sake of boilerplate, without any reason or effect.

File: .editorconfig [hunk 6]
```diff
@@ -0,0 +1,12 @@
+root = true
```
Comment (tumido) [inline .editorconfig:1]: Actually I have the same question as in the `.gitattributes` here. What does this bring? I don't see any reason for it, especially here in this repository where there's only a Dockerfile + service code around it. No `mdx` files. Things like indent make no sense here etc. What would change and what would be better if we merge this file?

Comment (schultzp2020) [reply .editorconfig:1]: The `.editorconfig` and `.gitattributes` are boilerplate and can definitely be reduced; however, the reason why I introuced them is because the default behavior on windows is to checkout files as crlf instead of lf. In the past this has caused problems for me.

**Comment by schultzp2020** [comment]
@tumido @sabre1041 can you take a look at this again?

**Review by tumido — COMMENTED** [review]
In general this LGTM. :+1: Thank you!

I think the @sabre1041 suggestion of keeping the `fix-permission` still stands... The dynamic nature of OCP uid/gid assignments should be addressed.

**Comment by schultzp2020** [comment]
> I think the @sabre1041 suggestion of keeping the `fix-permission` still stands... The dynamic nature of OCP uid/gid assignments should be addressed.

Whoops, I committed this locally but forgot to push it.

**Review by tumido — APPROVED** [review]
/lgtm
/approve

@sabre1041 PTAL

**Comment by schultzp2020** [comment]
@sabre1041 is this pr good to merge?

**Review by sabre1041 — APPROVED** [review]
LGTM
